stack_queue/[M]1249.cpp

```diff
@@ -0,0 +1,46 @@
+// Problem: 1249. Minimum Remove to Make Valid Parentheses - Medium
+// Link: https://leetcode.com/problems/minimum-remove-to-make-valid-parentheses/
+
+/* Nhận xét:
+- Nhận thấy, nếu các dấu ( và ) match với nhau thì đoạn phía trong nó là valid. Sau khi match toàn bộ các ( ) thì những dấu không valid còn lại sẽ có dạng )))(((.
+- Xóa những phần tử này đi ta sẽ có được string ta cần. Ta có thể dùng stack để matching các dấu với nhau.
+- Stack này cần lưu thêm index của dấu để có thể xóa đi sau khi duyệt xong.
+- Time complexity: O(n)
+- Space complexity: O(n) 
+*/
+
+class Solution {
+public:
+    string minRemoveToMakeValid(string s) {
+        int n = s.size();
+        stack<pair<char, int>> st;
+        for (int i = 0; i < n; ++i) {
+            char c = s[i];
+            if ('(' == c) {
+                st.push({c, i});
+            } else if (')' == c) {
+                if (!st.empty() && st.top().first == '(') {
+                    st.pop();
+                } else {
+                    st.push({c, i});
+                }
+            }
+        }
+        
+        stack<int> toRemove;
```

Comment:
ĐOạn từ dòng 30 này là hơi thừa. Có cách nào ngắn hơn ko? Hint: Cho 1 mảng removed[i] = true/false là kí tự thứ i có nên bị remove ko. Tính luôn mảng remove này trong phần tính stack bên trên luôn. Cuối cùng for 1 vòng cuối, nhặt các kí tự remove[i] = false là được.